### ISSUE TITLE
new relative perturbation system NaN and index errors

### ISSUE THREAD

**Issue by brycestx** [choderalab/perses#403]
I have been trying new perturbations from the JACS paper with the same CDK2 system in [relative_setup.py](https://github.com/choderalab/perses/blob/master/perses/distributed/relative_setup.py) and have received two errors which I believe to be related after running through a variety of perturbations (see below).

When defining molecule indices as 0 and 1 in [NonequilibriumFEPSetup](https://github.com/choderalab/perses/blob/master/perses/distributed/relative_setup.py#L679) function of relative_setup.py I received a NaN error:
```
[2017-09-21 18:44:42,096: ERROR/ForkPoolWorker-1] Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove
[2017-09-21 18:44:42,100: ERROR/ForkPoolWorker-1] Task perses.distributed.feptasks.run_equilibrium[c323162f-8b98-482f-b125-21a203802aa8] raised unexpected: IntegratorMoveError('Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove',)
Traceback (most recent call last):
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/celery/app/trace.py", line 374, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/celery/app/trace.py", line 629, in __protected_call__
    return self.run(*args, **kwargs)
  File "/home/ballen/github/perses/perses/distributed/feptasks.py", line 291, in run_equilibrium
    mc_move.apply(thermodynamic_state, sampler_state)
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/openmmtools-0.13.1-py3.6.egg/openmmtools/mcmc.py", line 1103, in apply
    super(LangevinDynamicsMove, self).apply(thermodynamic_state, sampler_state)
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/openmmtools-0.13.1-py3.6.egg/openmmtools/mcmc.py", line 703, in apply
    raise IntegratorMoveError(err_msg, self, context)
openmmtools.mcmc.IntegratorMoveError: Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove
[2017-09-21 18:44:47,255: INFO/MainProcess] missed heartbeat from worker1@alchemist
[2017-09-21 18:45:07,766: ERROR/MainProcess] Task handler raised error: <MaybeEncodingError: Error sending result: '"(1, <ExceptionInfo: IntegratorMoveError('Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove',)>, None)"'. Reason: ''PicklingError("Can\'t pickle <class \'openmmtools.mcmc.IntegratorMoveError\'>: it\'s not the same object as openmmtools.mcmc.IntegratorMoveError",)''.>
Traceback (most recent call last):
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/billiard/pool.py", line 362, in workloop
    put((READY, (job, i, result, inqW_fd)))
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/billiard/queues.py", line 366, in put
    self.send_payload(ForkingPickler.dumps(obj))
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/billiard/reduction.py", line 56, in dumps
    cls(buf, protocol).dump(obj)
billiard.pool.MaybeEncodingError: Error sending result: '"(1, <ExceptionInfo: IntegratorMoveError('Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove',)>, None)"'. Reason: ''PicklingError("Can\'t pickle <class \'openmmtools.mcmc.IntegratorMoveError\'>: it\'s not the same object as openmmtools.mcmc.IntegratorMoveError",)''.
[2017-09-21 18:45:07,771: INFO/MainProcess] missed heartbeat from worker2@alchemist
[2017-09-21 18:45:07,772: WARNING/MainProcess] Substantial drift from worker2@alchemist may mean clocks are out of sync.  Current drift is
25 seconds.  [orig: 2017-09-21 18:45:07.772171 recv: 2017-09-21 18:44:42.277738]

[2017-09-21 18:45:33,179: ERROR/ForkPoolWorker-1] Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove
[2017-09-21 18:45:33,182: ERROR/ForkPoolWorker-1] Task perses.distributed.feptasks.run_equilibrium[04b0caa4-51c1-4483-9888-2ff270746236] raised unexpected: IntegratorMoveError('Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove',)
Traceback (most recent call last):
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/celery/app/trace.py", line 374, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/celery/app/trace.py", line 629, in __protected_call__
    return self.run(*args, **kwargs)
  File "/home/ballen/github/perses/perses/distributed/feptasks.py", line 291, in run_equilibrium
    mc_move.apply(thermodynamic_state, sampler_state)
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/openmmtools-0.13.1-py3.6.egg/openmmtools/mcmc.py", line 1103, in apply
    super(LangevinDynamicsMove, self).apply(thermodynamic_state, sampler_state)
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/openmmtools-0.13.1-py3.6.egg/openmmtools/mcmc.py", line 703, in apply
    raise IntegratorMoveError(err_msg, self, context)
openmmtools.mcmc.IntegratorMoveError: Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove
[2017-09-21 18:45:37,270: INFO/MainProcess] missed heartbeat from worker1@alchemist
[2017-09-21 18:45:57,252: ERROR/MainProcess] Task handler raised error: <MaybeEncodingError: Error sending result: '"(1, <ExceptionInfo: IntegratorMoveError('Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove',)>, None)"'. Reason: ''PicklingError("Can\'t pickle <class \'openmmtools.mcmc.IntegratorMoveError\'>: it\'s not the same object as openmmtools.mcmc.IntegratorMoveError",)''.>
Traceback (most recent call last):
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/billiard/pool.py", line 362, in workloop
    put((READY, (job, i, result, inqW_fd)))
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/billiard/queues.py", line 366, in put
    self.send_payload(ForkingPickler.dumps(obj))
  File "/home/ballen/anaconda2/envs/oepython3/lib/python3.6/site-packages/billiard/reduction.py", line 56, in dumps
    cls(buf, protocol).dump(obj)
billiard.pool.MaybeEncodingError: Error sending result: '"(1, <ExceptionInfo: IntegratorMoveError('Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove',)>, None)"'. Reason: ''PicklingError("Can\'t pickle <class \'openmmtools.mcmc.IntegratorMoveError\'>: it\'s not the same object as openmmtools.mcmc.IntegratorMoveError",)''.
```
and when running the same [NonequilibriumFEPSetup](https://github.com/choderalab/perses/blob/master/perses/distributed/relative_setup.py#L679) function of relative_setup.py with molecule indices of 14 and 8 I receive a keyerror

```
Traceback (most recent call last):
  File "relative_setup.py", line 721, in <module>
    ne_fep.run(n_iterations=2)
  File "relative_setup.py", line 470, in run
    self._run_equilibrium()
  File "relative_setup.py", line 504, in _run_equilibrium
    positions_one = self._factory.new_positions(final_hybrid_positions_one)
  File "/home/ballen/github/perses/perses/annihilation/new_relative.py", line 1603, in new_positions
    new_positions[idx, :] = hybrid_positions[self._old_to_hybrid_map[idx], :]
KeyError: 2199

```

I am currently trying to debug the causes of these issues and will update with progress. I will update with serialized inputs as well if reproducibility is an issue.

Also if we determine that these issues are not related I can file separate a separate issue for each.

**Comment by jchodera** [comment]
In the first case, an `IntegratorMoveError` is thrown because a `NaN` was encountered and the `BaseIntegratorMove` was not configured to attempt to recover from this by retrying from the initial configuration (`move.n_restart_attempts = 0`). This could suggest something is problematic with the configuration, timestep, or alchemical Hamiltonian---all of which we should check out in detail---but we should probably be setting `move.n_restart_attempts = 5` [here](https://github.com/choderalab/perses/blob/master/perses/distributed/relative_setup.py#L389) to make the code minimally robust against NaNs.

We should also trap this Exception and make sure a `system.xml` and `state.xml` gets written to disk (ideally compressed on the fly via [zlib](https://docs.python.org/3/library/zlib.html)) to help us debug the cause of NaNs.

@pgrinaway will have to look into the second error in the atom mapping. I think it would be *incredibly* useful to have the mapper write out the atom mapping that it comes up with using the [atom labeling depiction toolkit](https://docs.eyesopen.com/toolkits/python/depicttk/moldisplay.html#atompropindex.py) from OpenEye---this would help us quickly spot any atom mapping issues.

**Comment by pgrinaway** [comment]
The second error looks like a typo:

`new_positions[idx, :] = hybrid_positions[self._old_to_hybrid_map[idx], :]`

should be

`new_positions[idx, :] = hybrid_positions[self._new_to_hybrid_map[idx], :]`

I'll take care of that first. Then, I'll tackle writing out system/state/integrator on error. I'll tackle the atom mapper thing last, since it is more involved.

Thanks for the issue!

**Comment by pgrinaway** [comment]
This was resolved in #404, and we haven't seen NaNs since. Closing, reopen if needed.
